File: python/patterns/src/prototype_patterns/behavioral/template_method.py
"""Template Method pattern example."""

from __future__ import annotations

from abc import ABC, abstractmethod


class DataPipeline(ABC):
    def run(self) -> list[str]:
        steps = [self.read(), self.transform(), self.write()]
        return steps

    @abstractmethod
    def read(self) -> str:
        raise NotImplementedError

    @abstractmethod
    def transform(self) -> str:
        raise NotImplementedError

    @abstractmethod
    def write(self) -> str:
        raise NotImplementedError


class CsvPipeline(DataPipeline):
    def read(self) -> str:
        return "read:csv"

    def transform(self) -> str:
        return "transform:normalize"

    def write(self) -> str:
        return "write:warehouse"


def demo() -> list[str]:
    return CsvPipeline().run()

File: nodejs/patterns/src/behavioral/template_method.js
class DataPipeline {
  run() {
    return [this.read(), this.transform(), this.write()];
  }

  read() {
    throw new Error("Not implemented");
  }

  transform() {
    throw new Error("Not implemented");
  }

  write() {
    throw new Error("Not implemented");
  }
}

class CsvPipeline extends DataPipeline {
  read() {
    return "read:csv";
  }

  transform() {
    return "transform:normalize";
  }

  write() {
    return "write:warehouse";
  }
}

function demo() {
  return new CsvPipeline().run();
}

module.exports = { CsvPipeline, demo };
